📖 Doc for updating  v1a1 provider to v1a2

akutz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
This patch includes developer documentation to assist with updating a v1alpha1 provider to v1alpha2.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
* Developer documentation for updating a v1alpha1 provider to v1alpha2
```

akutz commented:
/area kind/documentation

akutz commented:
/kind documentation

akutz commented on docs/developer/update-v1alpha1-provider-to-v1alpha2-infra-provider.md:
```diff
@@ -0,0 +1,188 @@
+
+### The spec and status types
+
+The `Spec` and `Status` types used to store configuration and status information are effectively the same in v1alpha2 as they were in v1alpha1:
```
TODO(@akutz) Mention that any kubeadm bootstrap fields/types should be removed

detiber replied:
Would it be good to call out that the kubeadm related configuration and CA certificate info has been extracted?

detiber commented on docs/developer/update-v1alpha1-provider-to-v1alpha2-infra-provider.md:
```diff
@@ -0,0 +1,188 @@
+
+### The infrastructure API group
+
+An infrastructure provider's CRDs should belong to the API group `infrastructure.cluster.x-k8s.io`.
```
This should probably also include a caveat that the `infrastructure.cluster.x-k8s.io` api group is only applicable for infrastructure providers that are housed under the `kubernetes-sigs` GitHub org.

Externally owned providers should use an appropriate API group for their ownership and would require additional RBAC rules to be configured and deployed for the common cluster-api components.

detiber commented on docs/developer/update-v1alpha1-provider-to-v1alpha2-infra-provider.md:
```diff
@@ -0,0 +1,188 @@
+
+### Kubebuilder
+
+While [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) v2 is available, the recommended approach for updating a CAPI provider to v1alpha2 is to stick with kubebuilder v1 during the update process and then reevaluate kubebuilder v2 after a successful migration to CAPI v1alpha2.
```
We may want to evaluate this down the line, I don't think we'll run into immediate issues, but at some point a migration from kubebuilder v1 to kubebuilder v2 will be needed to take advantage of webhooks, etc.

detiber commented on docs/developer/update-v1alpha1-provider-to-v1alpha2-infra-provider.md:
```diff
@@ -0,0 +1,188 @@
+| [`sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsprovider/v1alpha1.AWSMachineProviderSpec`](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/e6a57dc61826b8c7806eba22a513c9722c420754/pkg/apis/awsprovider/v1alpha1/awsmachineproviderconfig_types.go#L28-L97) | [`sigs.k8s.io/cluster-api-provider-aws/pkg/apis/infrastructure/v1alpha2.AWSMachineSpec`](https://github.com/kubernetes-sig ...
+| [`sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsprovider/v1alpha1.AWSMachineProviderStatus`](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/e6a57dc61826b8c7806eba22a513c9722c420754/pkg/apis/awsprovider/v1alpha1/awsmachineproviderstatus_types.go#L26-L44) | [`sigs.k8s.io/cluster-api-provider-aws/pkg/apis/infrastructure/v1alpha2.AWSMachineStatus`](https://github.com/kubernetes ...
+
+#### Provider cluster status fields
```
Can we call out specifically `Infrastructure Provider` here? I want to avoid conflation of the different types of providers going forward.

detiber commented on docs/developer/update-v1alpha1-provider-to-v1alpha2-infra-provider.md:
```diff
@@ -0,0 +1,188 @@
+
+### The provider cluster controller
+
+Instead of processing the CAPI `Cluster` resources like the old actuator model, the new provider cluster controller operates on the new provider `Cluster` CRD. However, the overall workflow should feel the same as the old cluster actuator. For example, take the AWS cluster controller's [reconcile function](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0a05127734a4fb955742b27c6e ...
```
This reads a bit confusing to me. instead of saying "new provider `Cluster` CRD", maybe say something along the lines of: "operates on a CRD that is defined by the infrastructure provider"?

Also, rather than "AWS cluster controller", it might be better to say "AWSCluster controller", so that it is clear that it is a controller for the `AWSCluster` type rather than a generic controller.

detiber commented on docs/developer/update-v1alpha1-provider-to-v1alpha2-infra-provider.md:
```diff
@@ -0,0 +1,188 @@
+
+Instead of processing the CAPI `Cluster` resources like the old actuator model, the new provider cluster controller operates on the new provider `Cluster` CRD. However, the overall workflow should feel the same as the old cluster actuator. For example, take the AWS cluster controller's [reconcile function](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0a05127734a4fb955742b27c6e ...
+
+1. Fetches the [AWS cluster resource](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0a05127734a4fb955742b27c6e326a65821851ce/pkg/controller/awscluster/awscluster_controller.go#L113-L121):
```
```suggestion
1. Fetches the [AWSCluster resource](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0a05127734a4fb955742b27c6e326a65821851ce/pkg/controller/awscluster/awscluster_controller.go#L113-L121):
```

detiber commented on docs/developer/update-v1alpha1-provider-to-v1alpha2-infra-provider.md:
```diff
@@ -0,0 +1,188 @@
+
+    If the AWS cluster resource does not have a corresponding CAPI cluster resource then the reconcile operation is requeued until such time the relationship is established.
+
+3. Uses a `defer` statement to [ensure the AWS and CAPI cluster resources are always patched](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0a05127734a4fb955742b27c6e326a65821851ce/pkg/controller/awscluster/awscluster_controller.go#L148-L153) back to the API server:
```
```suggestion
3. Uses a `defer` statement to [ensure the AWSCluster resource is always patched](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0a05127734a4fb955742b27c6e326a65821851ce/pkg/controller/awscluster/awscluster_controller.go#L148-L153) back to the API server:
```

Only the infrastructure provider defined resource should be updated by the infrastructure provider controllers, the related objects should only be used for reference

akutz commented:
> Thanks for starting on this, picked some nits related to terminology and some other minor comments.

Hi @detiber,

You're welcome. After @ncdc so generously walked me through v1alpha2, I figured I should pay it forward and write things down before I forget.

ykakarap commented:
@akutz I would like to help with this migration.

The first item is creating branches in the repo as details. I am afraid I will not have the needed permissions to do so.
Therefore I would like to start working on the second item: creating the provider's v1alpha2 resources. 

Please let me know if I can take it up.

ncdc commented on docs/developer/update-v1alpha1-provider-to-v1alpha2-infra-provider.md:
```diff
@@ -0,0 +1,188 @@
+    }
+    if cluster == nil {
+        logger.Info("Waiting for Cluster Controller to set OwnerRef on AWSCluster")
+        return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
```
We shouldn't need to requeue - we need the AWSCluster to get updated with an ownerRef to the Cluster, which means we'll process it again as part of a regular reconcile event (https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1037)

ncdc commented:
@akutz are you still interested & would you have time to port this over to the book content?

akutz commented:
Hi @ncdc,

Yes sir, I just haven't had time. I've got to prep for something this morning, but I should be able to address this before the end of the week.

ncdc commented:
Fantastic, thank you!

vincepri commented:
@akutz Will you have time to address the comments and port the content to the book?

vincepri commented:
@akutz Do you have time to address the comments?

akutz commented:
Hi @ncdc & @detiber,

I've addressed the feedback, but now I wonder the following:

1. Should this be added to the CAPI book? If so, where? @vincepri indicated there is a developer page, but it's more about getting started with the project as a developer, and less about what this document is attempting to capture.
2. Should I retarget this at `master`? It may be good to have for the historical record. It could also be closed as it's largely no longer relevant. It's up to y'all. It's my fault I took so long with this.

Thanks!

ncdc commented:
I would keep it in `master`, move it perhaps into `docs/book/src/providers/v1alpha1-to-v1alpha2.md`, and also backport to `release-0.2`.

ncdc commented:
Or we could just do this in `release-0.2` if that's preferable.

akutz commented:
Hi @ncdc,

I merged the two developer docs into the requested location.

ncdc commented:
Thanks @akutz!

/assign @chuckha 
as he wrote the initial doc

akutz commented:
Hi @chuckha,

I'm happy to add something about kubebuilder v2, but I'm wondering what you're thinking that isn't part of the kubebuilder section already?

chuckha commented:
Kubebuilder went through a lot of changes with the Makefile, main.go, directory structure. But that's actually all covered in the kubebuilder book! Maybe we should add a link out to this: https://book.kubebuilder.io/migration/v1vsv2.html and leave it at that? What do you think?

akutz commented:
Ah, I see. Yes, that sounds good. I'll make a note that if you move to kubebuilder v2, the changes are more than just code-based, and that a full review of the aforementioned link is a good idea.

akutz commented:
Hi @chuckha,

Done and done. Thank you for the suggestion.

chuckha commented:
/assign @ncdc 

for last review

ncdc commented:
/lgtm
